File: backend/scripts/snapshot_rule_runtime.py
import re

from pyspark.sql import functions as F
from pyspark.sql import types as T


TRANSFORM_OPERATIONS = {
    "cast",
    "copy",
    "default_value",
    "json_extract",
    "lowercase_trim",
    "mask",
    "null_guard",
    "parse_timestamp",
    "rename",
}
QUALITY_OPERATIONS = {"accepted_values", "not_null", "range", "regex"}
ROW_ID = "__asklake_rule_row_id"


class SnapshotRuleExecutionError(RuntimeError):
    def __init__(self, stage, rule, reason, *, transform=None, quality=None):
        super().__init__(f"{stage} rule {rule.get('id') or 'unknown'} failed: {reason}")
        self.failed_stage = stage
        self.quality = quality
        self.rule_id = str(rule.get("id") or "")
        self.transform = transform


def supports_snapshot_rules(rules):
    for rule in rules or []:
        if not rule or rule.get("enabled") is False:
            continue
        kind = str(rule.get("kind") or "")
        operation = str(rule.get("operation") or "")
        if kind == "transform" and operation in TRANSFORM_OPERATIONS:
            continue
        if kind == "quality" and operation in QUALITY_OPERATIONS:
            continue
        return False
    return True


def apply_snapshot_rules(frame, rules):
    canonical_rules = [rule for rule in (rules or []) if isinstance(rule, dict)]
    transforms = [rule for rule in canonical_rules if rule.get("kind") == "transform" and rule.get("enabled") is not False]
    quality_rules = [rule for rule in canonical_rules if rule.get("kind") == "quality" and rule.get("enabled") is not False]
    transform = {
        "appliedStepCount": 0,
        "configuredStepCount": len(transforms),
        "droppedCount": 0,
        "errorCount": 0,
        "quarantinedCount": 0,
        "setNullCount": 0,
        "warnCount": 0,
    }
    quality = {
        "blockingFailures": 0,
        "configuredRuleCount": len(quality_rules),
        "droppedCount": 0,
        "invalidRowCount": 0,
        "passRate": 100.0,
        "quarantinedCount": 0,
        "setNullCount": 0,
        "status": "pass",
        "summary": "Quality rules passed" if quality_rules else "No quality rules",
        "warnCount": 0,
    }
    current = frame.withColumn(ROW_ID, F.monotonically_increasing_id())
    quarantine = None

    for rule in transforms:
        output_name = _normalize_name(_first(rule.get("outputColumns")) or _first(rule.get("inputColumns")))
        if not output_name:
            continue
        row_count = current.count()
        expression, invalid, reason, source = _transform_expression(current, rule)
        invalid_count = current.filter(invalid).count()
        transform["appliedStepCount"] += row_count - invalid_count
        transform["errorCount"] += invalid_count
        action = _failure_action(rule)

        if invalid_count and action == "fail_batch":
            raise SnapshotRuleExecutionError(
                "transform",
                rule,
                reason,
                transform=transform,
                quality=quality,
            )
        if invalid_count and action == "quarantine":
            quarantine = _append_quarantine(
                quarantine,
                _quarantine_rows(current.filter(invalid), "transform", rule, reason),
            )
            transform["quarantinedCount"] += invalid_count
            current = current.filter(~invalid).withColumn(output_name, expression)
            continue
        if invalid_count and action == "drop_row":
            transform["droppedCount"] += invalid_count
            current = current.filter(~invalid).withColumn(output_name, expression)
            continue
        if invalid_count and action == "set_null":
            transform["setNullCount"] += invalid_count
            current = current.withColumn(output_name, F.when(invalid, F.lit(None)).otherwise(expression))
            continue
        if invalid_count:
            transform["warnCount"] += invalid_count
            current = current.withColumn(output_name, F.when(invalid, source).otherwise(expression))
            continue
        current = current.withColumn(output_name, expression)

    quality_input_count = current.count()
    quality["evaluatedRowCount"] = quality_input_count
    flag_names = []
    failure_reasons = []
    flagged = current
    for index, rule in enumerate(quality_rules):
        flag_name = f"__asklake_quality_failure_{index}"
        condition, reason = _quality_failure_condition(flagged, rule)
        flagged = flagged.withColumn(flag_name, condition)
        flag_names.append(flag_name)
        failure_reasons.append(reason)

    if flag_names:
        any_failure = F.col(flag_names[0])
        for flag_name in flag_names[1:]:
            any_failure = any_failure | F.col(flag_name)
        quality["invalidRowCount"] = flagged.filter(any_failure).select(ROW_ID).distinct().count()
    quality["passRate"] = (
        round(((quality_input_count - quality["invalidRowCount"]) / quality_input_count) * 100, 1)
        if quality_input_count
        else 100.0
    )

    blocking_rule = None
    blocking_reason = ""
    for rule, flag_name, reason in zip(quality_rules, flag_names, failure_reasons):
        if _failure_action(rule) != "fail_batch":
            continue
        failed_count = flagged.filter(F.col(flag_name)).count()
        quality["blockingFailures"] += failed_count
        if failed_count and blocking_rule is None:
            blocking_rule = rule
            blocking_reason = reason
    if blocking_rule is not None:
        quality["status"] = "fail"
        quality["summary"] = (
            f"Quality score {quality['passRate']}% - invalid rows {quality['invalidRowCount']} - "
            f"blocking {quality['blockingFailures']}"
        )
        raise SnapshotRuleExecutionError(
            "quality",
            blocking_rule,
            blocking_reason,
            transform=transform,
            quality=quality,
        )

    current = flagged
    for rule, flag_name, reason in zip(quality_rules, flag_names, failure_reasons):
        invalid = F.col(flag_name)
        invalid_count = current.filter(invalid).count()
        if not invalid_count:
            continue
        action = _failure_action(rule)
        if action == "quarantine":
            quarantine = _append_quarantine(
                quarantine,
                _quarantine_rows(current.filter(invalid), "quality", rule, reason),
            )
            quality["quarantinedCount"] += invalid_count
            current = current.filter(~invalid)
            continue
        if action == "drop_row":
            quality["droppedCount"] += invalid_count
            current = current.filter(~invalid)
            continue
        if action == "set_null":
            target = _resolve_column_name(current, _first(rule.get("inputColumns")))
            if target:
                current = current.withColumn(target, F.when(invalid, F.lit(None)).otherwise(F.col(_quote(target))))
            quality["setNullCount"] += invalid_count
            continue
        quality["warnCount"] += invalid_count

    if flag_names:
        current = current.drop(*flag_names)
    quality["status"] = "warn" if quality["invalidRowCount"] else "pass"
    quality["summary"] = (
        f"Quality score {quality['passRate']}% - invalid rows {quality['invalidRowCount']} - "
        f"dropped {quality['droppedCount']} - quarantined {quality['quarantinedCount']}"
        if quality_rules
        else "No quality rules"
    )
    return {
        "frame": current.drop(ROW_ID),
        "quality": quality,
        "quarantine": quarantine.drop(ROW_ID) if quarantine is not None else None,
        "transform": transform,
    }


def _transform_expression(frame, rule):
    input_name = _first(rule.get("inputColumns"))
    resolved = _resolve_column_name(frame, input_name)
    source = F.col(_quote(resolved)) if resolved else F.lit(None)
    operation = str(rule.get("operation") or "")
    parameters = rule.get("parameters") if isinstance(rule.get("parameters"), dict) else {}
    output_type = rule.get("outputType") or parameters.get("targetType")
    missing = _is_missing(source)

    if operation == "null_guard":
        expression, cast_invalid = _cast_column(source, output_type)
        return expression, missing | cast_invalid, "missing_required_value", source
    if operation == "default_value":
        selected = F.when(missing, F.lit(parameters.get("value"))).otherwise(source)
        expression, invalid = _cast_column(selected, output_type)
        return expression, invalid, "default_value_cast_failed", source
    if operation == "lowercase_trim":
        return F.lower(F.trim(source.cast("string"))), F.lit(False), "", source
    if operation == "json_extract":
        expression = F.get_json_object(_json_text(frame, resolved, source), str(parameters.get("path") or "$"))
        return expression, expression.isNull(), "json_path_not_found", source
    if operation == "mask":
        expression = F.regexp_replace(source.cast("string"), r"(\d{3})-?\d{4}-?(\d{4})", "$1-****-$2")
        return expression, F.lit(False), "", source
    if operation == "parse_timestamp":
        expression = F.to_timestamp(source.cast("string"))
        return expression, (~missing) & expression.isNull(), "timestamp_cast_failed", source
    if operation in {"cast", "copy", "rename"}:
        expression, invalid = _cast_column(source, output_type)
        return expression, invalid, "cast_failed", source
    return source, F.lit(True), "unsupported_transform_operation", source


def _cast_column(source, target_type):
    normalized = str(target_type or "").strip().lower()
    missing = _is_missing(source)
    if not normalized:
        return source, F.lit(False)
    text = F.trim(source.cast("string"))
    if "bool" in normalized:
        lowered = F.lower(text)
        expression = (
            F.when(lowered.isin("true", "1", "yes", "y"), F.lit(True))
            .when(lowered.isin("false", "0", "no", "n"), F.lit(False))
            .otherwise(F.lit(None).cast("boolean"))
        )
        return expression, (~missing) & expression.isNull()
    if any(token in normalized for token in ("integer", "int", "long", "bigint")):
        valid = text.rlike(r"^[+-]?\d+$")
        expression = F.when(valid, text.cast("long")).otherwise(F.lit(None).cast("long"))
        return expression, (~missing) & (~valid | expression.isNull())
    if any(token in normalized for token in ("float", "double", "decimal", "number", "numeric", "real")):
        valid = text.rlike(r"^[+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?$")
        expression = F.when(valid, text.cast("double")).otherwise(F.lit(None).cast("double"))
        return expression, (~missing) & (~valid | expression.isNull())
    if "timestamp" in normalized or "datetime" in normalized:
        expression = F.to_timestamp(text)
        return expression, (~missing) & expression.isNull()
    if normalized == "date":
        expression = F.to_date(text)
        return expression, (~missing) & expression.isNull()
    if any(token in normalized for token in ("json", "array", "struct", "map", "object")):
        return source.cast("string"), F.lit(False)
    return source.cast("string"), F.lit(False)


def _quality_failure_condition(frame, rule):
    target = _resolve_column_name(frame, _first(rule.get("inputColumns")))
    value = F.col(_quote(target)) if target else F.lit(None)
    operation = str(rule.get("operation") or "")
    parameters = rule.get("parameters") if isinstance(rule.get("parameters"), dict) else {}
    if operation == "not_null":
        return _is_missing(value), "missing_required_value"
    if operation == "range":
        text = F.trim(value.cast("string"))
        valid_number = text.rlike(r"^[+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?$")
        numeric = F.when(valid_number, text.cast("double")).otherwise(F.lit(None).cast("double"))
        condition = numeric.isNull()
        minimum = parameters.get("min")
        maximum = parameters.get("max")
        inclusive = parameters.get("inclusive") is not False
        if minimum not in (None, ""):
            try:
                minimum_value = float(minimum)
            except (TypeError, ValueError):
                return F.lit(True), "numeric_range_check_failed"
            condition = condition | (numeric < F.lit(minimum_value) if inclusive else numeric <= F.lit(minimum_value))
        if maximum not in (None, ""):
            try:
                maximum_value = float(maximum)
            except (TypeError, ValueError):
                return F.lit(True), "numeric_range_check_failed"
            condition = condition | (numeric > F.lit(maximum_value) if inclusive else numeric >= F.lit(maximum_value))
        return condition, "numeric_range_check_failed"
    if operation == "regex":
        pattern = str(parameters.get("pattern") or "")
        try:
            re.compile(pattern)
        except re.error:
            return F.lit(True), "invalid_regex_pattern"
        return value.isNull() | ~value.cast("string").rlike(pattern), "regex_match_failed"
    if operation == "accepted_values":
        values = [str(item) for item in parameters.get("values", [])] if isinstance(parameters.get("values"), list) else []
        return value.isNull() | ~value.cast("string").isin(*values), "value_not_accepted"
    return F.lit(True), "unsupported_quality_operation"


def _quarantine_rows(frame, stage, rule, reason):
    visible_columns = [column for column in frame.columns if not column.startswith("__asklake_quality_failure_")]
    record_columns = [F.col(_quote(column)).alias(column) for column in visible_columns if column != ROW_ID]
    event_id = _resolve_column_name(frame, "event_id")
    return frame.select(
        F.col(ROW_ID),
        (F.col(_quote(event_id)).cast("string") if event_id else F.lit("")).alias("event_id"),
        F.to_json(F.struct(*record_columns)).alias("record"),
        F.lit(reason).alias("reason"),
        F.lit(str(rule.get("id") or "")).alias("ruleId"),
        F.lit(stage).alias("stage"),
        F.lit(str(_first(rule.get("inputColumns")) or _first(rule.get("outputColumns")) or "")).alias("targetColumn"),
    )


def _append_quarantine(current, next_frame):
    return next_frame if current is None else current.unionByName(next_frame)


def _failure_action(rule):
    if rule.get("onError") == "fail_batch":
        return "fail_batch"
    if rule.get("onError") == "quarantine":
        return "quarantine"
    if rule.get("failureDisposition") == "drop_row":
        return "drop_row"
    if rule.get("failureDisposition") == "set_null":
        return "set_null"
    return "keep"


def _is_missing(column):
    return column.isNull() | (F.length(F.trim(column.cast("string"))) == 0)


def _resolve_column_name(frame, name):
    text = str(name or "")
    if text in frame.columns:
        return text
    normalized = _normalize_name(text)
    return normalized if normalized in frame.columns else ""


def _json_text(frame, resolved, source):
    if not resolved:
        return source.cast("string")
    data_type = frame.schema[resolved].dataType
    if isinstance(data_type, (T.ArrayType, T.MapType, T.StructType)):
        return F.to_json(source)
    return source.cast("string")


def _normalize_name(value):
    text = re.sub(r"[^0-9A-Za-z_]+", "_", str(value or "").strip()).strip("_")
    return text


def _quote(name):
    return f"`{str(name).replace('`', '``')}`"


def _first(values):
    return values[0] if isinstance(values, list) and values else ""

File: backend/src/snapshotRuleRuntime.mjs
const transformOperations = new Set([
  "cast",
  "copy",
  "default_value",
  "json_extract",
  "lowercase_trim",
  "mask",
  "null_guard",
  "parse_timestamp",
  "rename",
]);
const qualityOperations = new Set(["accepted_values", "not_null", "range", "regex"]);

export class SnapshotRuleExecutionError extends Error {
  constructor(stage, rule, reason, context = {}) {
    super(`${stage} rule ${rule?.id || "unknown"} failed: ${reason}`);
    this.name = "SnapshotRuleExecutionError";
    this.failedStage = stage;
    this.quality = context.quality;
    this.ruleId = String(rule?.id || "");
    this.transform = context.transform;
  }
}

export function supportsSnapshotRules(rules) {
  return list(rules)
    .filter((rule) => rule?.enabled !== false)
    .every((rule) => (
      rule?.kind === "transform"
        ? transformOperations.has(String(rule.operation || ""))
        : rule?.kind === "quality" && qualityOperations.has(String(rule.operation || ""))
    ));
}

export function applySnapshotRules(records, rules) {
  const canonicalRules = list(rules).filter((rule) => rule && typeof rule === "object");
  const transforms = canonicalRules.filter((rule) => rule.kind === "transform" && rule.enabled !== false);
  const qualityRules = canonicalRules.filter((rule) => rule.kind === "quality" && rule.enabled !== false);
  const transform = {
    appliedStepCount: 0,
    configuredStepCount: transforms.length,
    droppedCount: 0,
    errorCount: 0,
    quarantinedCount: 0,
    setNullCount: 0,
    warnCount: 0,
  };
  const quality = {
    blockingFailures: 0,
    configuredRuleCount: qualityRules.length,
    droppedCount: 0,
    invalidRowCount: 0,
    passRate: 100,
    quarantinedCount: 0,
    setNullCount: 0,
    status: "pass",
    summary: qualityRules.length > 0 ? "Quality rules passed" : "No quality rules",
    warnCount: 0,
  };
  const quarantined = [];
  let transformed = list(records).map((record) => structuredClone(record));

  for (const rule of transforms) {
    const output = String(rule.outputColumns?.[0] || rule.inputColumns?.[0] || "");
    if (!output) continue;
    const outcomes = transformed.map((record) => {
      try {
        return { record, value: applyTransformRule(record, rule) };
      } catch (error) {
        return { error, record };
      }
    });
    const failures = outcomes.filter((outcome) => outcome.error);
    transform.appliedStepCount += outcomes.length - failures.length;
    transform.errorCount += failures.length;
    const action = failureAction(rule);
    if (failures.length > 0 && action === "fail_batch") {
      const reason = failures[0].error?.code || failures[0].error?.message || String(failures[0].error);
      throw new SnapshotRuleExecutionError("transform", rule, reason, { quality, transform });
    }

    const next = [];
    for (const outcome of outcomes) {
      if (!outcome.error) {
        setRecordValue(outcome.record, output, outcome.value);
        next.push(outcome.record);
        continue;
      }
      const reason = outcome.error?.code || outcome.error?.message || String(outcome.error);
      if (action === "drop_row") {
        transform.droppedCount += 1;
        continue;
      }
      if (action === "quarantine") {
        quarantined.push(quarantineEntry(outcome.record, "transform", rule, reason));
        transform.quarantinedCount += 1;
        continue;
      }
      if (action === "set_null") {
        setRecordValue(outcome.record, output, null);
        transform.setNullCount += 1;
      } else {
        setRecordValue(outcome.record, output, getRecordValue(outcome.record, rule.inputColumns?.[0]));
        transform.warnCount += 1;
      }
      next.push(outcome.record);
    }
    transformed = next;
  }

  const evaluated = transformed.map((record) => ({
    failures: qualityRules.map((rule) => ({
      reason: qualityFailureReason(getRecordValue(record, rule.inputColumns?.[0]), rule),
      rule,
    })),
    record,
  }));
  const qualityInputCount = transformed.length;
  quality.evaluatedRowCount = qualityInputCount;
  quality.invalidRowCount = evaluated.filter(({ failures }) => failures.some(({ reason }) => reason)).length;
  quality.passRate = qualityInputCount
    ? Number((((qualityInputCount - quality.invalidRowCount) / qualityInputCount) * 100).toFixed(1))
    : 100;

  let blockingFailure;
  for (const rule of qualityRules) {
    if (failureAction(rule) !== "fail_batch") continue;
    const failed = evaluated.filter(({ failures }) => failures.some((item) => item.rule === rule && item.reason));
    quality.blockingFailures += failed.length;
    if (!blockingFailure && failed.length > 0) {
      blockingFailure = failed[0].failures.find((item) => item.rule === rule && item.reason);
    }
  }
  if (blockingFailure) {
    quality.status = "fail";
    quality.summary = `Quality score ${quality.passRate}% - invalid rows ${quality.invalidRowCount} - blocking ${quality.blockingFailures}`;
    throw new SnapshotRuleExecutionError("quality", blockingFailure.rule, blockingFailure.reason, { quality, transform });
  }

  const output = [];
  for (const { failures, record } of evaluated) {
    const activeFailures = failures.filter((item) => item.reason);
    if (activeFailures.length === 0) {
      output.push(record);
      continue;
    }

    let discarded = false;
    for (const { reason, rule } of activeFailures) {
      const action = failureAction(rule);
      if (action === "drop_row") {
        quality.droppedCount += 1;
        discarded = true;
        break;
      }
      if (action === "quarantine") {
        quarantined.push(quarantineEntry(record, "quality", rule, reason));
        quality.quarantinedCount += 1;
        discarded = true;
        break;
      }
      if (action === "set_null") {
        setRecordValue(record, rule.inputColumns?.[0], null);
        quality.setNullCount += 1;
      } else {
        quality.warnCount += 1;
      }
    }
    if (!discarded) output.push(record);
  }

  quality.status = quality.invalidRowCount > 0 ? "warn" : "pass";
  quality.summary = qualityRules.length > 0
    ? `Quality score ${quality.passRate}% - invalid rows ${quality.invalidRowCount} - dropped ${quality.droppedCount} - quarantined ${quality.quarantinedCount}`
    : "No quality rules";

  return { quality, quarantined, records: output, transform };
}

function applyTransformRule(record, rule) {
  const input = getRecordValue(record, rule.inputColumns?.[0]);
  const operation = String(rule.operation || "");
  const parameters = object(rule.parameters);
  const outputType = rule.outputType;

  if (operation === "null_guard") {
    if (isMissing(input)) throw ruleError("missing_required_value");
    return castValue(input, outputType);
  }
  if (operation === "default_value") {
    return castValue(isMissing(input) ? parameters.value : input, outputType);
  }
  if (operation === "lowercase_trim") {
    return input === null || input === undefined ? null : String(input).trim().toLowerCase();
  }
  if (operation === "json_extract") {
    const value = readJsonPath(input, parameters.path);
    if (value === undefined) throw ruleError("json_path_not_found");
    return castValue(value, outputType || "String");
  }
  if (operation === "mask") {
    return input === null || input === undefined ? null : maskPhoneNumber(String(input));
  }
  if (operation === "parse_timestamp") {
    return castValue(input, "Timestamp");
  }
  if (operation === "cast" || operation === "copy" || operation === "rename") {
    return castValue(input, outputType || parameters.targetType);
  }
  throw ruleError("unsupported_transform_operation");
}

function castValue(value, targetType) {
  if (isMissing(value)) return null;
  const type = String(targetType || "").trim().toLowerCase();
  if (!type) return value;
  if (type.includes("bool")) {
    const normalized = String(value).trim().toLowerCase();
    if (["true", "1", "yes", "y"].includes(normalized)) return true;
    if (["false", "0", "no", "n"].includes(normalized)) return false;
    throw ruleError("boolean_cast_failed");
  }
  if (["integer", "int", "long", "bigint"].some((token) => type.includes(token))) {
    const normalized = String(value).trim();
    if (!/^[+-]?\d+$/.test(normalized)) throw ruleError("integer_cast_failed");
    const numeric = Number(normalized);
    if (!Number.isSafeInteger(numeric)) throw ruleError("integer_cast_failed");
    return numeric;
  }
  if (["float", "double", "decimal", "number", "numeric", "real"].some((token) => type.includes(token))) {
    const numeric = Number(String(value).trim());
    if (!Number.isFinite(numeric)) throw ruleError("numeric_cast_failed");
    return numeric;
  }
  if (type.includes("timestamp") || type.includes("datetime")) {
    const timestamp = new Date(String(value).trim());
    if (Number.isNaN(timestamp.getTime())) throw ruleError("timestamp_cast_failed");
    return timestamp.toISOString().replace(".000Z", "Z");
  }
  if (type === "date") {
    const timestamp = new Date(String(value).trim());
    if (Number.isNaN(timestamp.getTime())) throw ruleError("date_cast_failed");
    return timestamp.toISOString().slice(0, 10);
  }
  if (["json", "array", "struct", "map", "object"].some((token) => type.includes(token))) {
    if (value && typeof value === "object") return structuredClone(value);
    try {
      return JSON.parse(String(value));
    } catch {
      throw ruleError("json_cast_failed");
    }
  }
  return value && typeof value === "object" ? JSON.stringify(value) : String(value);
}

function qualityFailureReason(value, rule) {
  const operation = String(rule.operation || "");
  const parameters = object(rule.parameters);
  if (operation === "not_null") return isMissing(value) ? "missing_required_value" : "";
  if (operation === "range") {
    if (isMissing(value)) return "numeric_range_check_failed";
    const numeric = Number(String(value).trim());
    const minimum = parameters.min === undefined || parameters.min === null || parameters.min === ""
      ? Number.NEGATIVE_INFINITY
      : Number(parameters.min);
    const maximum = parameters.max === undefined || parameters.max === null || parameters.max === ""
      ? Number.POSITIVE_INFINITY
      : Number(parameters.max);
    const inclusive = parameters.inclusive !== false;
    const valid = Number.isFinite(numeric)
      && !Number.isNaN(minimum)
      && !Number.isNaN(maximum)
      && (inclusive ? numeric >= minimum && numeric <= maximum : numeric > minimum && numeric < maximum);
    return valid ? "" : "numeric_range_check_failed";
  }
  if (operation === "regex") {
    try {
      return new RegExp(String(parameters.pattern || "")).test(String(value ?? "")) ? "" : "regex_match_failed";
    } catch {
      return "invalid_regex_pattern";
    }
  }
  if (operation === "accepted_values") {
    const values = Array.isArray(parameters.values) ? parameters.values.map(String) : [];
    return values.includes(String(value ?? "")) ? "" : "value_not_accepted";
  }
  return "unsupported_quality_operation";
}

function failureAction(rule) {
  if (rule.onError === "fail_batch") return "fail_batch";
  if (rule.onError === "quarantine") return "quarantine";
  if (rule.failureDisposition === "drop_row") return "drop_row";
  if (rule.failureDisposition === "set_null") return "set_null";
  return "keep";
}

function quarantineEntry(record, stage, rule, reason) {
  return {
    event_id: String(getRecordValue(record, "event_id") ?? ""),
    reason,
    record: structuredClone(record),
    ruleId: String(rule.id || ""),
    stage,
    targetColumn: String(rule.inputColumns?.[0] || rule.outputColumns?.[0] || ""),
  };
}

function getRecordValue(record, field) {
  const pathParts = String(field || "").split(".").filter(Boolean);
  if (pathParts.length === 0) return undefined;
  if (Object.hasOwn(record, field)) return record[field];
  let value = record;
  for (const part of pathParts) {
    if (!value || typeof value !== "object") return undefined;
    value = value[part];
  }
  if (value !== undefined) return value;
  if (record.raw && typeof record.raw === "object") {
    const rawField = String(field).replace(/^raw[_.]/, "");
    return record.raw[rawField] ?? record.raw[field];
  }
  return undefined;
}

function setRecordValue(record, field, value) {
  const parts = String(field || "").split(".").filter(Boolean);
  if (parts.length === 0) return;
  if (parts.length === 1) {
    record[parts[0]] = value;
    return;
  }
  let target = record;
  for (const part of parts.slice(0, -1)) {
    if (!target[part] || typeof target[part] !== "object") target[part] = {};
    target = target[part];
  }
  target[parts.at(-1)] = value;
}

function readJsonPath(input, expression) {
  if (input === null || input === undefined) throw ruleError("json_input_missing");
  const source = typeof input === "string" ? JSON.parse(input) : input;
  const pathParts = String(expression || "$").replace(/^\$\.?/, "").split(".").filter(Boolean);
  return pathParts.reduce((value, part) => value?.[part], source);
}

function maskPhoneNumber(value) {
  return value.replace(/(\d{3})-?\d{4}-?(\d{4})/, "$1-****-$2");
}

function isMissing(value) {
  return value === null || value === undefined || (typeof value === "string" && value.trim() === "");
}

function ruleError(code) {
  const error = new Error(code);
  error.code = code;
  return error;
}

function object(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? { ...value } : {};
}

function list(value) {
  return Array.isArray(value) ? value : [];
}
